File: backend/app/api/sources.py
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.orm import Session

from ..core.vocab import MaterialKind, MaterialStatus
from ..domain.models import Material, MaterialSource
from ..jobs.runner import JobRunner
from ..services.content.sources import SourcesError, SourcesService
from ..services.platform.profiles import ensure_default_profile
from .deps import get_session

router = APIRouter(prefix="/sources", tags=["sources"])


class SourceBrowseOut(BaseModel):
    source_id: int
    label: str
    path: str
    subdir: str
    missing_target: bool
    enabled: bool
    scan_interval_sec: int | None
    last_scan_error: str | None
    last_scanned_at: str | None
    subdirs: list["SourceSubdirOut"]
    materials: list["SourceMaterialOut"]
    uningested: list["UningestedFileOut"]


class SourceSubdirOut(BaseModel):
    name: str


class SourceMaterialOut(BaseModel):
    id: int
    title: str
    kind: MaterialKind
    status: MaterialStatus
    filename: str
    relpath: str


class UningestedFileOut(BaseModel):
    name: str
    relpath: str
    size_bytes: int
    mtime: float


SourceBrowseOut.model_rebuild()


class SourceIn(BaseModel):
    label: str = Field(min_length=1, max_length=200)
    path: str = Field(min_length=1, max_length=1000)
    recursive: bool = True
    include_globs: list[str] | None = None
    course_id: int
    scan_interval_sec: int | None = Field(default=None, ge=15)


class SourceOut(BaseModel):
    id: int
    label: str
    path: str
    recursive: bool
    include_globs: list[str] | None
    course_id: int
    enabled: bool
    scan_interval_sec: int | None = None
    last_scan_error: str | None = None
    material_count: int
    last_scanned_at: str | None


def _source_out(session: Session, source: MaterialSource, material_count: int = 0) -> SourceOut:
    return SourceOut(
        id=source.id,
        label=source.label,
        path=source.path,
        recursive=source.recursive,
        include_globs=source.include_globs,
        course_id=source.course_id,
        enabled=source.enabled,
        scan_interval_sec=source.scan_interval_sec,
        last_scan_error=source.last_scan_error,
        material_count=material_count,
        last_scanned_at=(
            source.last_scanned_at.isoformat() if source.last_scanned_at else None
        ),
    )


class ScanResult(BaseModel):
    stats: dict[str, int]
    queued_jobs: int


@router.get("", response_model=list[SourceOut])
def list_sources(
    request: Request, session: Session = Depends(get_session)
) -> list[SourceOut]:
    profile = ensure_default_profile(session)
    service = SourcesService(session, request.app.state.settings.blobs_dir)
    result: list[SourceOut] = []
    for source in service.list_sources(profile.id):
        count = len(
            session.scalars(
                select(Material.id).where(Material.source_id == source.id)
            ).all()
        )
        result.append(_source_out(session, source, count))
    return result


@router.post("", response_model=SourceOut, status_code=201)
def add_source(
    body: SourceIn,
    request: Request,
    session: Session = Depends(get_session),
) -> SourceOut:
    profile = ensure_default_profile(session)
    service = SourcesService(session, request.app.state.settings.blobs_dir)
    try:
        source = service.create_source(
            profile.id,
            label=body.label,
            path=body.path,
            recursive=body.recursive,
            include_globs=body.include_globs,
            course_id=body.course_id,
            scan_interval_sec=body.scan_interval_sec,
        )
    except SourcesError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error
    session.commit()
    return _source_out(session, source)


@router.delete("/{source_id}", status_code=204)
def delete_source(
    source_id: int,
    request: Request,
    session: Session = Depends(get_session),
) -> None:
    profile = ensure_default_profile(session)
    service = SourcesService(session, request.app.state.settings.blobs_dir)
    if not service.delete_source(profile.id, source_id):
        raise HTTPException(status_code=404, detail="source not found")
    session.commit()


@router.get("/{source_id}/browse", response_model=SourceBrowseOut)
def browse_source(
    source_id: int,
    request: Request,
    subdir: str = "",
    session: Session = Depends(get_session),
) -> dict[str, Any]:
    profile = ensure_default_profile(session)
    service = SourcesService(session, request.app.state.settings.blobs_dir)
    try:
        return service.browse(profile.id, source_id, subdir)
    except SourcesError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error


class IngestFileIn(BaseModel):
    relpath: str = Field(min_length=1, max_length=2000)


class IngestFileOut(BaseModel):
    material_id: int
    job_id: int | None
    deduped: bool


@router.post("/{source_id}/ingest", status_code=201, response_model=IngestFileOut)
def ingest_source_file(
    source_id: int,
    body: IngestFileIn,
    request: Request,
    session: Session = Depends(get_session),
) -> dict[str, Any]:
    profile = ensure_default_profile(session)
    service = SourcesService(session, request.app.state.settings.blobs_dir)
    try:
        material, deduped = service.ingest_file(profile.id, source_id, body.relpath)
    except SourcesError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error
    job_id: int | None = None
    if not deduped:
        job = JobRunner.enqueue(session, "ingest", {"material_id": material.id})
        job_id = job.id
    session.commit()
    request.app.state.jobs.wake()
    return {"material_id": material.id, "job_id": job_id, "deduped": deduped}


class RelinkIn(BaseModel):
    path: str = Field(min_length=1, max_length=1000)


@router.patch("/{source_id}", response_model=SourceOut)
def relink_source(
    source_id: int,
    body: RelinkIn,
    request: Request,
    session: Session = Depends(get_session),
) -> SourceOut:
    profile = ensure_default_profile(session)
    service = SourcesService(session, request.app.state.settings.blobs_dir)
    try:
        source = service.relink(profile.id, source_id, body.path)
    except SourcesError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error
    session.commit()
    return _source_out(session, source)


@router.post("/{source_id}/reveal", status_code=204)
def reveal_source(
    source_id: int,
    request: Request,
    session: Session = Depends(get_session),
) -> None:
    profile = ensure_default_profile(session)
    service = SourcesService(session, request.app.state.settings.blobs_dir)
    try:
        service.reveal(profile.id, source_id)
    except SourcesError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error


def _enqueue_ingests(session: Session, source_id: int) -> int:
    pending = session.scalars(
        select(Material).where(
            Material.source_id == source_id, Material.status == "pending"
        )
    ).all()
    for material in pending:
        JobRunner.enqueue(session, "ingest", {"material_id": material.id})
    return len(pending)


class ScanAllOut(BaseModel):
    scanned: int
    results: dict[str, dict[str, int]]


@router.post("/scan-all", response_model=ScanAllOut)
def scan_all_sources(
    request: Request,
    session: Session = Depends(get_session),
) -> dict[str, Any]:
    del session
    scheduler = getattr(request.app.state, "scans", None)
    if scheduler is None:
        raise HTTPException(status_code=503, detail="scan scheduler not running")
    results = scheduler.scan_all(force=True)
    return {"scanned": len(results), "results": results}


@router.post("/{source_id}/scan", response_model=ScanResult)
def scan_source(
    source_id: int,
    request: Request,
    session: Session = Depends(get_session),
) -> ScanResult:
    profile = ensure_default_profile(session)
    service = SourcesService(session, request.app.state.settings.blobs_dir)
    try:
        stats = service.scan(profile.id, source_id)
    except SourcesError as error:
        session.rollback()
        failure = SourcesService(session, request.app.state.settings.blobs_dir)
        try:
            failing = failure._get(profile.id, source_id)
            failing.last_scan_error = str(error)
            session.commit()
        except SourcesError:
            session.rollback()
        raise HTTPException(status_code=422, detail=str(error)) from error
    queued = _enqueue_ingests(session, source_id)
    session.commit()
    request.app.state.jobs.wake()
    return ScanResult(stats=stats, queued_jobs=queued)

File: frontend/src/lib/api/sources.ts
import { json, apiFetch } from './client'
import type { ImportResult } from './quiz'

export interface BrowseEntry {
  name: string
}

export interface BrowseMaterial {
  id: number
  title: string
  kind: string
  status: string
  filename: string
  relpath: string
}

export interface BrowsePending {
  name: string
  relpath: string
  size_bytes: number
  mtime: number
}

export interface BrowseResult {
  source_id: number
  label: string
  path: string
  subdir: string
  missing_target: boolean
  enabled: boolean
  scan_interval_sec: number | null
  last_scan_error: string | null
  last_scanned_at: string | null
  subdirs: BrowseEntry[]
  materials: BrowseMaterial[]
  uningested: BrowsePending[]
}

export async function browseSource(
  sourceId: number,
  subdir = ''
): Promise<BrowseResult> {
  const params = subdir ? `?subdir=${encodeURIComponent(subdir)}` : ''
  const response = await apiFetch(`/api/v1/sources/${sourceId}/browse${params}`)
  return json<BrowseResult>(response)
}

export async function ingestSourceFile(
  sourceId: number,
  relpath: string
): Promise<{ material_id: number; job_id: number | null; deduped: boolean }> {
  const response = await apiFetch(`/api/v1/sources/${sourceId}/ingest`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ relpath }),
  })
  return json<{ material_id: number; job_id: number | null; deduped: boolean }>(response)
}

export async function relinkSource(sourceId: number, path: string): Promise<LinkedSource> {
  const response = await apiFetch(`/api/v1/sources/${sourceId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ path }),
  })
  return json<LinkedSource>(response)
}

export async function revealSource(sourceId: number): Promise<void> {
  const response = await apiFetch(`/api/v1/sources/${sourceId}/reveal`, {
    method: 'POST',
  })
  if (!response.ok && response.status !== 204) {
    throw new Error(`reveal failed: ${response.status}`)
  }
}

export interface LinkedSource {
  id: number
  label: string
  path: string
  recursive: boolean
  include_globs: string[] | null
  course_id: number
  enabled: boolean
  material_count: number
  last_scanned_at: string | null
}

export interface ScanResult {
  stats: { new: number; updated: number; unchanged: number; missing: number }
  queued_jobs: number
}

export async function listSources(): Promise<LinkedSource[]> {
  const response = await apiFetch('/api/v1/sources')
  return json<LinkedSource[]>(response)
}

export async function addSource(body: {
  label: string
  path: string
  course_id: number
  recursive?: boolean
}): Promise<LinkedSource> {
  const response = await apiFetch('/api/v1/sources', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  return json<LinkedSource>(response)
}

export async function scanSource(sourceId: number): Promise<ScanResult> {
  const response = await apiFetch(`/api/v1/sources/${sourceId}/scan`, { method: 'POST' })
  return json<ScanResult>(response)
}

export async function deleteSource(sourceId: number): Promise<void> {
  const response = await apiFetch(`/api/v1/sources/${sourceId}`, { method: 'DELETE' })
  if (!response.ok && response.status !== 204) {
    throw new Error(`delete failed: ${response.status}`)
  }
}

export interface InboxEntryInfo {
  filename: string
  kind: string
  title: string
  ok: boolean
  problems: string[]
  question_count: number
}

export async function scanInbox(): Promise<InboxEntryInfo[]> {
  const response = await apiFetch('/api/v1/quiz/inbox')
  return json<InboxEntryInfo[]>(response)
}

export async function inboxPath(): Promise<string> {
  const response = await apiFetch('/api/v1/quiz/inbox/path')
  return (await json<{ path: string }>(response)).path
}

export async function importInboxFile(
  filename: string,
  courseId: number
): Promise<ImportResult> {
  const params = new URLSearchParams({ course_id: String(courseId) })
  const response = await apiFetch(
    `/api/v1/quiz/inbox/${encodeURIComponent(filename)}/import?${params.toString()}`,
    { method: 'POST' }
  )
  return json<ImportResult>(response)
}
